tests/publiccloud/migration.pm

```diff
@@ -21,42 +21,117 @@ use utils;
 use publiccloud::utils;
 
 our $target_version = get_required_var('TARGET_VERSION');
-our $not_clean_vm = get_var('PUBLIC_CLOUD_NO_CLEANUP_ON_FAILURE');
+our $not_clean_vm   = get_var('PUBLIC_CLOUD_NO_CLEANUP_ON_FAILURE');
 
 sub run {
     my ($self, $args) = @_;
     select_serial_terminal();
     my $provider = $args->{my_provider};
-    my $instance = $provider->create_instance();
-    $instance->wait_for_guestregister();
+    my $instance = $provider->create_instance(check_guestregister => 0);
     registercloudguest($instance) if is_byos();
 
     register_addons_in_pc($instance);
-    $instance->ssh_assert_script_run('sudo zypper ref; sudo zypper -n up', timeout => 300);
-
     record_info('Target Version', $target_version);
 
     sleep 90;    # wait for a bit for zypper to be available
 
-    record_info('installs SLE15-Migration and suse-migration-sle15-activation');
+    #Refresh and Update
+    $instance->ssh_assert_script_run( 'sudo zypper ref; sudo zypper -n up',
+        timeout => 300 );
+
+    #Download unpublished DMS package and install rpm
+    my $url              = get_var('PUBLIC_CLOUD_DMS_IMAGE_LOCATION');
+    my $package          = get_var('PUBLIC_CLOUD_DMS_PACKAGE');
+    my $pc_url           = get_var('PUBLIC_CLOUD_DMS_PC_LOCATION');
+    my $pc_pkg           = get_var('PUBLIC_CLOUD_DMS_PC_PACKAGE');
+    my $act_url          = get_var('PUBLIC_CLOUD_DMS_ACT_LOCATION');
+    my $act_pkg          = get_var('PUBLIC_CLOUD_DMS_ACT_PACKAGE');
+    my $dms_repo_key_url = get_var('PUBLIC_CLOUD_DMS_REPO_KEY_URL');
+    my $dms_repo_key     = get_var('PUBLIC_CLOUD_DMS_REPO_KEY');
+    my $dms_rpm          = "$url" . "$package";
+    my $act_rpm          = 'N/A';
+    my $pc_rpm           = 'N/A';
+    my ( $v1, $dms_rpm_name ) = split /\//, $package;
+    my $act_out = `curl -s $act_url`;
+    $act_rpm = $1 if $act_out =~ /($act_pkg.*?rpm)/g;
```

Comment:
isn't this duplication of what you already has in `PUBLIC_CLOUD_DMS_ACT_PACKAGE`  variable ?

Reply:
so why do we need this variable at all ? Instead of passing two variables `PUBLIC_CLOUD_DMS_PC_LOCATION` and `PUBLIC_CLOUD_DMS_PC_PACKAGE`  just pass `PUBLIC_CLOUD_DMS_PC_URL` will contains `https://download.suse.de/ibs/Devel:/PubCloud:/CrossCloud/SLE_12_SP4/noarch/suse-migration-pre-checks-2.0.36-1.1.noarch.rpm`  and you can easily cut package name from it . 

I don't point of passing two variables and than doing some things on top here with them . Or we need to improve logic in trigger_public_cloud and pass normal package name which not need any additional actions here or we passing single variable and extracting package name here . What you currently suggesting is doing same job ( extract package name ) two times . First in trigger_public_cloud script and than second time here

Reply:
first let me create the repo and fix the install packages and get this PR merge

and then I will refactor the code both trigger_public_cloud and migration code. we can fix this exact package name and complete URL in trigger_public_cloud script itself.